examples/matrix_multiplier/tests/test_matrix_multiplier.py

```diff
@@ -0,0 +1,163 @@
+import math
+import os
+from random import getrandbits
+
+import cocotb
+from cocotb.binary import BinaryValue
+from cocotb.clock import Clock
+from cocotb.monitors import BusMonitor
+from cocotb.regression import TestFactory
+from cocotb.scoreboard import Scoreboard
+from cocotb.triggers import RisingEdge, ReadOnly
+
+NUM_SAMPLES = int(os.environ.get('NUM_SAMPLES', 3000))
+DATA_WIDTH = int(os.environ['DATA_WIDTH'])
+A_ROWS = int(os.environ['A_ROWS'])
+B_COLUMNS = int(os.environ['B_COLUMNS'])
+A_COLUMNS_B_ROWS = int(os.environ['A_COLUMNS_B_ROWS'])
+
+MULTI_DIMENSIONAL_ARRAYS = not cocotb.SIM_NAME.lower().startswith(("icarus", "ghdl"))
+
+if MULTI_DIMENSIONAL_ARRAYS:
+    def get_matrix_value(signal, rows, cols):
+        return signal.value
+
+    def set_matrix_value(signal, matrix_value):
+        signal <= matrix_value
+else:
+    def get_matrix_value(signal, rows, cols):
+        return [
+            [
+                signal[(i * cols) + j].value
+                for j in range(cols)
+            ]
+            for i in range(rows)
+        ]
+
+    def set_matrix_value(signal, matrix_value):
+        for i, row in enumerate(matrix_value):
+            for j, value in enumerate(row):
+                signal[(i * len(row)) + j] <= value
```

Comment:
One more idea, not sure if it's worthwhile:
```suggestion
class UnpackedMatrixSignal:
    def __init__(self, signal, rows, cols):
        self.signal = signal
        self.rows = rows
        self.cols = cols

    @property
    def value(self):
        return [
            [
                self.signal[(i * self.cols) + j].value
                for j in range(self.cols)
            ]
            for i in range(self.rows)
        ]

    @value.setter
    def value(signal, matrix_value):
        for i, row in enumerate(matrix_value):
            for j, value in enumerate(row):
                self.signal[(i * len(row)) + j].value = value

if MULTI_DIMENSIONAL_ARRAYS:
    matrix_signal = lambda x, rows, cols: x
else:
    matrix_signal = UnpackedMatrixSignal
```

Reply:
I've used wrapper classes around signals before, but there might be something simpler.

Remove `set_matrix_value` and just generate 1D array since we recover values from the signals anyway on `i_valid` assertion to calculate expected result.
```python3
if MULTI_DIMENSIONAL_ARRAYS:
    def gen_matrix(func, rows, cols):
        return [[func(DATA_WIDTH) for col in range(cols)] for row in range(rows)]
else:
    def gen_matrix(func, rows, cols):
        return [func(DATA_WIDTH) for col in range(cols) for row in range(rows)]
```
Not sure about `get_matrix value`. It might be nice to use the `value` getter on the top matrix handle, then "unflatten" to 2D in Python.